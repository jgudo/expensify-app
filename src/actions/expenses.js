import uuid from 'uuid';

//ADD_EXPENSE
export const addExpense = ( {description = '', note = '', amount = 0, createdAt = 0} = {} ) => ({
    type: 'ADD_EXPENSE',
    expenses: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }

});
//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//REMOVE_EXPENSE

export const removeExpense = ( {id} ) => ({
    type: 'REMOVE_EXPENSE',
    id
});