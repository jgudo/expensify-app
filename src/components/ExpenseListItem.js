import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({id,description, amount, createdAt}) =>  (
    <Link to={`/edit/${id}`} className="list-item">
        <div className="list-item__description">
            <h3 className="list-item__title">{description}</h3>
            <span>{moment(createdAt).format('MMMM Do YYYY')}</span>
        </div>
        <div>
            <h3> ₱{numeral(amount / 100).format('₱0,0.00')}</h3>
        </div>
    </Link>
);

export default ExpenseListItem;
//export default ExpenseListItem;