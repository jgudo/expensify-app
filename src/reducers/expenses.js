//Reducer 
const expensesReducer = (state = [], action) => {
    switch(action.type) {
        case 'ADD_EXPENSE': 
            return [...state, action.expenses];
        
        case 'REMOVE_EXPENSE': 
            return state.filter( (exp) => exp.id !== action.id );
        
        case 'EDIT_EXPENSE': 
            return state.map((expense) => {
                if(expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                };
            });
        default: 
            return state;
    }
}

export default expensesReducer;