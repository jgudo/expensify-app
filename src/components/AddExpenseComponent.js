import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpenseComponent extends React.Component {
    onSubmit = (expense) => {
        this.props.onSubmit(expense);
        //props.dispatch(addExpense(expense));
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                 <h1>Add New Expense Form</h1>
                 <ExpenseForm 
                     onSubmit={this.onSubmit}
                 />
            </div>
         
         );
    }
}


const mapDispatchToProps = (dispatch) => ({
    onSubmit: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpenseComponent);