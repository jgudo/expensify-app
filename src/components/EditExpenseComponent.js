import React from 'react';
import {connect} from 'react-redux';
import  ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses';

export class EditExpenseComponent extends React.Component {

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
        
    }
    // onRemove = () => {
    //     this.props.removeExpense({id: this.props.expense.id});
    //     this.props.history.push('/');
    //    
    // }

    render() {
        return(
            <div>
            <ExpenseForm
                expense={this.props.expense}
                onSubmit={this.onSubmit}
            />
            <button 
                onClick={() => {
                this.props.removeExpense({id: this.props.expense.id});
                this.props.history.push('/');
                //props.dispatch(removeExpense({id:props.expense.id}));
            }}
            >
             Remove
        
             </button>
        </div>
        );
    }
}
const mapDispatchToProps = (dispatch) => ({
    editExpense: (id,expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
});

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find( (expense) => expense.id === props.match.params.id)
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpenseComponent);
