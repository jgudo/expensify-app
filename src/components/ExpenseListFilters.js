import React from 'react';
import { connect } from 'react-redux';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
        console.log('STARTDATE:', startDate, ' END DATE: ', endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused }));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }

    onSelectChange = (e) => {
        if(e.target.value === 'date') {
            this.props.sortByDate();
        } else if(e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    }

    render() {
        return (
            <div>
                <input type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}

                />
                <select onChange={this.onSelectChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    startDateId="start_date"
                    endDate={this.props.filters.endDate}
                    endDateId="end_date"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={ () => false}
                />
                
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))  
});

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);