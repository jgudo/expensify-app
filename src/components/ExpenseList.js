import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {props.expenses.length === 0 ?  (
            <p>No Expenses List</p>
        ) : (
            props.expenses.map( (item) => {
                return <ExpenseListItem key={item.id} {...item}/>
            })
        )}
     
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};
export default connect(mapStateToProps)(ExpenseList);

