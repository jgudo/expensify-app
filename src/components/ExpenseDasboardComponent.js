import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardComponent = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardComponent;