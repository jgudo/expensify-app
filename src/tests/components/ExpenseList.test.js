import React from 'react';
import {shallow} from 'enzyme';
import { ExpenseList} from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';


test('Should render expense list with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render message if not expense passed', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});