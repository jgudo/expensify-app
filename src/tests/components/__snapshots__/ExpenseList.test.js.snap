// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense list with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={21424}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={135237}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={1241}
    createdAt={345600000}
    description="Water Bill"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`Should render message if not expense passed 1`] = `
<div>
  <p>
    No Expenses List
  </p>
</div>
`;
